### ISSUE TITLE
More control over autoScroll

### ISSUE THREAD

**Issue by spectraldoy** [hello-pangea/dnd#431]
Hi,

This library is awesome. It has greatly empowered my projects. However, I would like to make a feature request: disabling autoScroll dynamically. For my current use case, I _need_ to be able to drag elements without scrolling the container of my Droppables. Trying to dynamically set `overflow` CSS parameter of the container that contains my Droppables does not work -- there seems to be some sort of override. But, there are many use cases where disabling autoScroll, particularly while dragging something, would be useful or convenient.

As far as I can tell (and I would love to be corrected), this library does not support disabling autoScroll, or controlling the speed of autoScroll, or customizing the margins at which it begins, etc. I am requesting that this feature be added - the ability for the user to set autoScrolling thresholds, and to dynamically enable or disable it.

The original `react-beautiful-dnd` library also had many similar requests. The following are some examples:
- https://github.com/atlassian/react-beautiful-dnd/issues/1902
- https://github.com/atlassian/react-beautiful-dnd/issues/460

The second issue in particular is interesting, due to this comment: https://github.com/atlassian/react-beautiful-dnd/issues/460#issuecomment-581066434. That comment contains a link to [this repository](https://github.com/mjuopperi/react-beautiful-dnd/commits/optional-autoscroll), where the @mjuopperi forked the repository, and was able to add an autoScroll option to the `DragDropContext` actually without much fuss.

If given the assurance that this feature would be accepted if I were to implement it, I'll go figure it out and set up a pull request ASAP. Actually, regardless of whether or not it will be accepted as a contribution to this library, I'll probably go tinker with it. But please let me know if this feature request would be accepted.

Thank you.

**Comment by spectraldoy** [comment]
Following up on this: I've started working on my own copy of the repository to try and add an `autoScroll` prop to the `DragDropContext`. Can someone please give me some directions / pointers on how to build the project without getting any errors?

**Comment by Xhale1** [comment]
Hello! Id love to review and accept a PR which adds the option to disable auto scroll.

As for building the project, are you hitting any specific error? I can review and update our docs tonight too.

**Comment by spectraldoy** [comment]
@Xhale1 Thank you very much. I'm not too familiar with building and releasing libraries, so I wasn't exactly sure where to start. I will document everything I tried in this message, but the most important thing is that I don't know how the build the project properly so that it can be imported as a package. I would appreciate if you told me your general workflow for building the project and running tests so that I can follow a proper sequence of steps without having to come up with heuristics that simply get things to compile without errors.

I also apologize for the length of this message. I'm simply documenting everything I tried. You do not necessarily need to know everything - just that a general workflow for building the project would very much be appreciated.

Basically, in order to test the library while making changes, my overall attempts can be described as follows:
1. `create-react-app --template typescript` to build a test app that I could use to test the DragDropContext with an autoScroll prop
2. `git clone https://github.com/hello-pangea/dnd/` from within the `src/` folder of the created React App
3. Make my changes to the clone of `dnd` inside `src/`, i.e., `src/dnd/`

However, I realized that when importing from `dnd`, it would default to searching for a `dist/` directory inside `dnd/`, i.e. it would search for `src/dnd/dist`. So, my first attempt at editing `dnd` was to build it.

### Building `dnd`

I ran `pnpm install` within the `src/dnd` directory to install all the dependencies, and then `pnpm build` to build the project. However, once this was done, it created an error that I just could not figure out how to solve:

```
Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:

1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app See for tips about how to debug and fix this problem.
```

It also printed out something along the lines of `TypeError: Cannot read properties of null (reading 'useMemo')`. I apologize, I currently am unable to give you the exact error because it would involve making a lot of changes to my current test repo.

However, it printed out this same error for several React hooks, `useMemo`, `useState`, etc. From the line numbers where it said these errors were occurring, I discovered that they were coming from functions that overwrote these hooks using `resolveDispatcher` (which I believe comes from React):

<img width="317" alt="image" src="https://user-images.githubusercontent.com/41366802/201466808-b7ff61a9-71f5-49fd-9861-4203291715ad.png">

(in `bundle.js` while Inspect Elementing from `localhost:3000` after running `npm start`)

I know this error was not caused by the test code I wrote, because I did not have `useMemo` nor `useEffect` in my test code. What might have been causing this, as outlined in that "Invalid Hook call message," is having multiple versions of React - one in the root directory created by Create React App, and one inside `src/dnd/`. However, removing React from `src/dnd` did not fix this error. I did not try removing it from the root directory.

I did not know how to fix this, so I tried the following.

### Making `dnd` "source code"

I moved `package.json` and basically every other config file (like `babel.config.js`) or directory (like `cypress/`) out of `src/dnd` into the root directory created by Create React App. I removed the `dist` folder from `src/dnd`, deleted `src/dnd/.git` and inside my `src/App.tsx` file, I have the following line:

```js
import { DragDropContext } from "./dnd/src";
```

I then ran `npm install` and then `npm start`. VSCode does not throw any errors. But when I tried `npm start`, it gave me several "missing dependency" errors due to packages that `npm install` had not installed (some examples: redux, react-redux, memoize-one, use-memo-one). After installing these, I ran into several typescript and module resolution errors. In order to solve these, I deleted the `test/` and `stories/` directories (which were throwing most of the errors, even though I was not importing from them), added `/* eslint-disable */` to several files, such as my `App.tsx`, and prettierignored everything in `/src`. The errors due to prettier were coming from `dnd`, not from the test code that I added, so I wasn't sure what to do, and this works as a short-term solution. There was also an error with `src/dnd/src/view/drag-drop-context/use-startup-validation.ts`, where `peerDependencies` was being imported from `package.json`, and the compiler was throwing an error, complaining that this was bad practice. I simply commented out the associated lines.

Now it compiles without errors, but as you can tell, only after some surgery.

 I'll update you with my progress as I go along. My current plan is to make changes under this test framework, make a second clone of the repository, and then simply add my changes to that one without deleting things and moving files around, then making a pull request. For that, I would really appreciate a proper list of steps for building and testing the project.

Thank you.

**Comment by spectraldoy** [comment]
Update: I have added an `enableAutoScroll` prop to `DragDropContext`, and it seems to work. I need to write some proper tests for it though. I'll get a PR up as soon as possible, but I would like to investigate adding some more control over autoScroll, like the maximum speed, and the thresholds.

**Comment by spectraldoy** [comment]
Update: PR is up: https://github.com/hello-pangea/dnd/pull/434

**Comment by Xhale1** [comment]
Merged in comit 82779377a4faf93a91f9cfca0a1be0bb35278ee6, PR #434
